### PR TITLE
fix(DatePicker/CalendarMonth): better range styling when some dates are disabled

### LINKED ISSUE #9933: Bug - Datepicker - End date validator lacks support for "isSameOrBefore" case
**Describe the problem**
The PatternFly DatePicker component exhibits unexpected behavior when using the `toValidator` function in a date range selection scenario. The issue arises when the validator condition checks if the end date is greater than the start date (`yyyy-MM-DD > yyyy-MM-DD`). This results in the selected styles for the end date range not being applied correctly.

**Affected Components**
- DatePicker

**How do you reproduce the problem?**
Steps to reproduce:
1. Set up a DatePicker component to select a date range.
2. Implement a `toValidator` function similar to this:
   ```javascript
   const toValidator = (date) => {
       return isValidDate(from) && yyyyMMddFormat(date) > yyyyMMddFormat(from)
         ? ''
         : 'The "to" date must be after the "from" date';
   };
   ```
3. Select a start date at the end of the month.
4. Open the end date datepicker and observe that the selected styles for the end date range are not applied.

You can reproduce this issue within the [PF datepicker demo example](https://www.patternfly.org/components/date-and-time/date-and-time-picker/#date-and-time-range-picker).

**Expected behavior**
The expected behavior is that when selecting a start date at the end of the month, the end date datepicker should show the correct beginning range with selected styles applied. 

**Is this issue blocking you?**
Not blocking. 

**Screenshots**
Screenshot 2024-01-03 at 4 04 10 PM
Screenshot 2024-01-03 at 4 05 50 PM

**Environment**
- OS: [e.g. macOS]
- Browser: [e.g. Chrome]
- Version: [e.g. 90.0]

**Product and Release Date**
(Product details and intended release date for the feature using this component)

**Additional Information**
Upon changing the `toValidator` function condition from `>=` to `>`, the DatePicker component's end date range start selected styles do not apply as intended. It appears the PatternFly DatePicker requires an end date range that starts with the start date for proper functioning.

### PR DESCRIPTION
**Pull request by adamviktora**
Closes #9933

- shows styling on a range even if some of the dates of the range (including `rangeStart`) may be disabled with the `validators` prop
- tries to autofocus the `dateProp` first (rangeEnd) over rangeStart. If no dateProp is given, tries to autofocus the `rangeStart` over today

Related to discussion on the working session - I decided **not** to attempt to find a closest focusable date, neither to add the prop to customize which month should be open. It might be useful only in rare use case like having a validator that disables large amount of dates (and it could complicate things more than be useful). - but let me know if it is something we might want to add

### CODE REVIEW

File: packages/react-core/src/components/DatePicker/DatePicker.tsx [hunk 1]
```diff
@@ -231,12 +231,25 @@ const DatePickerBase = (
   const createFocusSelectorString = (modifierClass: string) =>
     `.${calendarMonthStyles.calendarMonthDatesCell}.${modifierClass} .${calendarMonthStyles.calendarMonthDate}`;
   const focusSelectorForSelectedDate = createFocusSelectorString(calendarMonthStyles.modifiers.selected);
+  const focusSelectorForSelectedEndRangeDate = createFocusSelectorString(
+    `${calendarMonthStyles.modifiers.selected}.${calendarMonthStyles.modifiers.endRange}`
+  );
   const focusSelectorForUnselectedDate = createFocusSelectorString(calendarMonthStyles.modifiers.current);
 
+  const getElementToFocus = () => {
```
Comment (nicolethoen) [inline packages/react-core/src/components/DatePicker/DatePicker.tsx:239]: Could you add a comment here to just explain the purpose of this function? I think this all looks good but I'm also finding the complexity of all this stuff pretty hard to follow 😂

File: packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx [hunk 2]
```diff
@@ -412,8 +412,8 @@ export const CalendarMonth = ({
                       isAdjacentMonth && styles.modifiers.adjacentMonth,
                       isToday && styles.modifiers.current,
                       (isSelected || isRangeStart) && styles.modifiers.selected,
-                      !isValid && styles.modifiers.disabled,
-                      (isInRange || isRangeStart || isRangeEnd) && styles.modifiers.inRange,
+                      !isValid && !(isInRange || isRangeStart || isRangeEnd || isSelected) && styles.modifiers.disabled,
```
[the remark this answers is not in the document]
Comment (adamviktora) [reply packages/react-core/src/components/CalendarMonth/CalendarMonth.tsx:415]: I think this is due to the docs building error resolved in https://github.com/patternfly/patternfly-react/pull/10416, it looks fine locally.

**Review by thatblindgeye — APPROVED** [review]
Latest build looks good when using the validator mentioned in the issue

**Review by nicolethoen — APPROVED** [review]
